src/test/java/ejercicio/banco/search/ToolSearchEngineTest.java

```diff
@@ -1,5 +1,88 @@
 package ejercicio.banco.search;
 
+import ejercicio.banco.service.AccountServiceImpl;
+import ejercicio.banco.service.BankServiceImpl;
+import ejercicio.banco.service.PaymentServiceImpl;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import java.io.File;
+import java.util.List;
+
+import static ejercicio.banco.dto.DataType.ACCOUNT;
+import static ejercicio.banco.dto.DataType.BANK;
+import static ejercicio.banco.dto.DataType.PAYMENT;
+import static junit.framework.TestCase.assertEquals;
+import static org.mockito.Mockito.verify;
+
+@RunWith(MockitoJUnitRunner.class)
 public class ToolSearchEngineTest {
-    //TODO: Tests
+
+    private static final String BANK_NAME = "SHADYBANK";
+    private static final String ACCOUNT_NAME = "Pedro";
+    private static final String PAYMENT_NAME = "5";
+    private static final String INVALID_NAME = "NOTHING";
+    private static final String BANK_FILE = String.join(File.separator, "src", "main", "resources", "csv", "randomBank.csv");
+    private static final String ACCOUNT_FILE = String.join(File.separator, "src", "main", "resources", "csv", "randomAccounts.csv");
+    private static final String PAYMENT_FILE = String.join(File.separator, "src", "main", "resources", "csv", "randomPayments.csv");
+
+    @Mock
+    private BankServiceImpl bankService;
+    @Mock
+    private AccountServiceImpl accountService;
+    @Mock
+    private PaymentServiceImpl paymentService;
+
+    @InjectMocks
+    private ToolSearchEngine searchEngine;
+
+    @Rule
+    public ExpectedException expectedEx = ExpectedException.none();
+
+    @Test
+    public void testSearchThrowsExceptionWhenDataTypeIsNull() {
+        expectedEx.expect(IllegalArgumentException.class);
+        searchEngine.search(null, null);
+        expectedEx.expectMessage("Type must not be null");
+    }
+
+    @Test
+    public void testSearchThrowsExceptionWhenStuffIsNull() {
+        expectedEx.expect(NullPointerException.class);
+        searchEngine.search(BANK, null);
+        expectedEx.expectMessage("Searching Null String");
+    }
+
+    //TODO: Seems like when returning Collections, the size of the array increases
+
+    @Test
+    public void testSearchReturnsEmptyArrayWhenStuffDoesNotMatch() {
+        List<Object> results =  searchEngine.search(BANK, INVALID_NAME);
+        assertEquals(1,results.size());
+    }
+
+    //TODO: None of the verify tests seem to work
```

Comment:
Yup, I just realized where the issue is. Take a look to the `ToolSearchEngine` implementation in the switch-else.

The services are being instantiated each time the search engine is called, which is wrong :)

I mean this code:

```
BankService bankService = new BankServiceImpl(new CsvBankRepository());
[..]
AccountService accountService = new AccountServiceImpl(new CsvAccountRepository());
[..]
List<Payment> payments = ((PaymentServiceImpl) paymentService).findPayments(file, stuff);
```

Instead of doing that, try to pass these services in the constructor. Something like:

```
public ToolSearchEngine(BankService bankService, AccountService accountService, PaymentService paymentService) {
}
```